File: backend/src/main/kotlin/com/aicoach/service/DashboardService.kt
package com.aicoach.service

import com.aicoach.domain.dto.DashboardSummaryDto
import com.aicoach.domain.dto.WeekAdherenceDto
import com.aicoach.repository.StravaActivityRepository
import com.aicoach.repository.TrainingPlanRepository
import com.aicoach.repository.WeeklyBlockRepository
import org.springframework.stereotype.Service
import java.time.LocalDate

@Service
class DashboardService(
    private val planRepository: TrainingPlanRepository,
    private val weeklyBlockRepository: WeeklyBlockRepository,
    private val activityRepository: StravaActivityRepository
) {
    fun getSummary(): DashboardSummaryDto {
        val plan = planRepository.findFirstByOrderByIdAsc().orElse(null)
            ?: return DashboardSummaryDto(emptyList(), null, 0.0, 0.0)

        val weeks = weeklyBlockRepository.findByTrainingPlanIdOrderByWeekNumberAsc(plan.id)
        val today = LocalDate.now()

        val weekAdherences = weeks.map { week ->
            val activities = activityRepository
                .findByActivityDateBetweenOrderByActivityDateAsc(week.startDate, week.endDate)
                .filter { it.sportType in listOf("Run", "TrailRun", "VirtualRun") }

            val actualKm = activities.sumOf { it.distanceKm }
            val actualVertM = activities.sumOf { it.totalElevationM?.toDouble() ?: 0.0 }
            val plannedKm = week.plannedKm ?: 0.0
            val plannedVertM = week.plannedVertM ?: 0

            val adherence = if (plannedKm > 0) (actualKm / plannedKm * 100).coerceAtMost(200.0) else 0.0
            val isCurrentWeek = !today.isBefore(week.startDate) && !today.isAfter(week.endDate)
            val isFutureWeek = today.isBefore(week.startDate)

            WeekAdherenceDto(
                weekNumber = week.weekNumber,
                phase = week.phase,
                startDate = week.startDate,
                endDate = week.endDate,
                plannedKm = plannedKm,
                actualKm = actualKm,
                plannedVertM = plannedVertM,
                actualVertM = actualVertM,
                adherencePercent = adherence,
                activityCount = activities.size,
                isCurrentWeek = isCurrentWeek,
                isFutureWeek = isFutureWeek
            )
        }

        val currentWeekNumber = weekAdherences.find { it.isCurrentWeek }?.weekNumber
        val totalPlannedKm = weekAdherences.filter { !it.isFutureWeek }.sumOf { it.plannedKm }
        val totalActualKm = weekAdherences.sumOf { it.actualKm }

        return DashboardSummaryDto(
            weeks = weekAdherences,
            currentWeekNumber = currentWeekNumber,
            totalPlannedKm = totalPlannedKm,
            totalActualKm = totalActualKm
        )
    }

    fun getSummaryForAthlete(athleteId: Long): DashboardSummaryDto {
        val plan = planRepository.findByAthleteId(athleteId).orElse(null)
            ?: return DashboardSummaryDto(emptyList(), null, 0.0, 0.0)

        val weeks = weeklyBlockRepository.findByTrainingPlanIdOrderByWeekNumberAsc(plan.id)
        val today = LocalDate.now()

        val weekAdherences = weeks.map { week ->
            val activities = activityRepository
                .findByInternalAthleteIdAndActivityDateBetweenOrderByActivityDateAsc(
                    athleteId, week.startDate, week.endDate
                )
                .filter { it.sportType in listOf("Run", "TrailRun", "VirtualRun") }

            val actualKm = activities.sumOf { it.distanceKm }
            val actualVertM = activities.sumOf { it.totalElevationM?.toDouble() ?: 0.0 }
            val plannedKm = week.plannedKm ?: 0.0
            val plannedVertM = week.plannedVertM ?: 0

            val adherence = if (plannedKm > 0) (actualKm / plannedKm * 100).coerceAtMost(200.0) else 0.0
            val isCurrentWeek = !today.isBefore(week.startDate) && !today.isAfter(week.endDate)
            val isFutureWeek = today.isBefore(week.startDate)

            WeekAdherenceDto(
                weekNumber = week.weekNumber,
                phase = week.phase,
                startDate = week.startDate,
                endDate = week.endDate,
                plannedKm = plannedKm,
                actualKm = actualKm,
                plannedVertM = plannedVertM,
                actualVertM = actualVertM,
                adherencePercent = adherence,
                activityCount = activities.size,
                isCurrentWeek = isCurrentWeek,
                isFutureWeek = isFutureWeek
            )
        }

        val currentWeekNumber = weekAdherences.find { it.isCurrentWeek }?.weekNumber
        val totalPlannedKm = weekAdherences.filter { !it.isFutureWeek }.sumOf { it.plannedKm }
        val totalActualKm = weekAdherences.sumOf { it.actualKm }

        return DashboardSummaryDto(
            weeks = weekAdherences,
            currentWeekNumber = currentWeekNumber,
            totalPlannedKm = totalPlannedKm,
            totalActualKm = totalActualKm
        )
    }
}

File: server/src/services/dashboardService.ts
import { db } from '../db/client.js'
import { trainingPlans, weeklyBlocks } from '../db/schema.js'
import { eq } from 'drizzle-orm'
import { getActivitiesByDateRange } from './stravaActivityService.js'
import type { DashboardSummaryDto, WeekAdherenceDto } from '../types/index.js'

async function buildDashboard(
  planId: number,
  athleteId: number,
): Promise<DashboardSummaryDto> {
  const weeks = (await db
    .select()
    .from(weeklyBlocks)
    .where(eq(weeklyBlocks.trainingPlanId, planId))
    .all())
    .sort((a, b) => a.weekNumber - b.weekNumber)

  const today = new Date().toISOString().substring(0, 10)

  const weekAdherences: WeekAdherenceDto[] = await Promise.all(weeks.map(async (week) => {
    const activities = await getActivitiesByDateRange(athleteId, week.startDate, week.endDate)

    const actualKm =
      activities.reduce((sum, a) => sum + (a.distanceM ? a.distanceM / 1000 : 0), 0)
    const actualVertM = activities.reduce(
      (sum, a) => sum + (a.totalElevationM ?? 0),
      0
    )
    const plannedKm = week.plannedKm ?? 0
    const plannedVertM = week.plannedVertM ?? 0

    const adherence = plannedKm > 0
      ? Math.min((actualKm / plannedKm) * 100, 200)
      : 0

    const isCurrentWeek = today >= week.startDate && today <= week.endDate
    const isFutureWeek = today < week.startDate

    return {
      weekNumber: week.weekNumber,
      phase: week.phase ?? null,
      startDate: week.startDate,
      endDate: week.endDate,
      plannedKm,
      actualKm: Math.round(actualKm * 100) / 100,
      plannedVertM,
      actualVertM: Math.round(actualVertM * 10) / 10,
      adherencePercent: Math.round(adherence * 10) / 10,
      activityCount: activities.length,
      isCurrentWeek,
      isFutureWeek,
    }
  }))

  const currentWeekNumber =
    weekAdherences.find((w) => w.isCurrentWeek)?.weekNumber ?? null
  const totalPlannedKm = weekAdherences
    .filter((w) => !w.isFutureWeek)
    .reduce((sum, w) => sum + w.plannedKm, 0)
  const totalActualKm = weekAdherences.reduce((sum, w) => sum + w.actualKm, 0)

  return {
    weeks: weekAdherences,
    currentWeekNumber,
    totalPlannedKm: Math.round(totalPlannedKm * 100) / 100,
    totalActualKm: Math.round(totalActualKm * 100) / 100,
  }
}

export async function getDashboardSummary(): Promise<DashboardSummaryDto> {
  const plans = await db.select().from(trainingPlans).all()
  const plan = plans.sort((a, b) => a.id - b.id)[0]
  if (!plan) return { weeks: [], currentWeekNumber: null, totalPlannedKm: 0, totalActualKm: 0 }

  return buildDashboard(plan.id, plan.athleteId ?? 0)
}

export async function getDashboardSummaryForAthlete(athleteId: number): Promise<DashboardSummaryDto> {
  const plan = await db
    .select()
    .from(trainingPlans)
    .where(eq(trainingPlans.athleteId, athleteId))
    .get()
  if (!plan) return { weeks: [], currentWeekNumber: null, totalPlannedKm: 0, totalActualKm: 0 }

  return buildDashboard(plan.id, athleteId)
}
